[Bootloader] Add basic jump to application support

gpappasv opened the pull request.

gpappasv committed:
Add readme.md file under projects folder

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

gpappasv committed:
Move compile_commands and .map file under build folder

This commit introduces a modification on the CMakeLists files of the
app/bootloader projects so as to move the map and compile_commands files
under build folder, after building the projects
Also, a build script is created under scripts to ease the build process.

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

gpappasv committed:
Add jump to application basic functionality on bootloader

This commit is about adding the jump to application support on the
bootloader project.
Also, the available flash size of the bootloader is now reduced to 32kb.
UART2 is also being used by the bootloader application now.

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

gpappasv committed:
Adjust the app linker script flash section

Reduce the flash section from 512KB to 480KB
Flash section now starts from 0x8008000 address

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

gpappasv committed:
Add initial bootloader architecture diagram (under docs)

This commit is about creating the bootloader architecture in a draw.io
diagram. Note: this diagram will change a lot over time.
Also, under this commit, the project structure is expanded based on the
architecture diagram.

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

xristosioan commented on projects/app/boards/stm32f401re/STM32F401RETx_FLASH.ld:
```diff
@@ -62,7 +62,7 @@ _Min_Stack_Size = 0x400; /* required amount of stack */
 {
 RAM (xrw)      : ORIGIN = 0x20000000, LENGTH = 96K
-FLASH (rx)      : ORIGIN = 0x8000000, LENGTH = 512K
+FLASH (rx)      : ORIGIN = 0x8008000, LENGTH = 480K /* 512K */
```
Define symbols for start address and length for both bootloader and applications, in order to make them readable from bootloader's source code. This change will make jumping to application dynamic, based on the value inside the linker file.

gpappasv replied:
The 0x8008000 information exists on the linker script of the actual application. From the bootloader side we dont have access to this and we cannot provide this information to the application.
1) Bootloader will eventually (try to) jump to a specific flash address (0x8008000).
2) Application will adjust its linker script to 0x8008000, to align with the bootloader requirement.
I dont see how defining the symbol on the bootloader linker script, or on app linker script will ease the integration.

xristosioan replied:
The whole point is to avoid the 0x8008000 completely, by just exposing the address as APP_CODE_BANK1_START_ADDRESS or smth. Of course this symbol will exist at both linker scripts. The user must adjust it at both linker script files if he wishes to.

gpappasv replied:
Let's leave it for a next PR

gpappasv committed:
Clean-up on bootloader project

This commit is about cleaning up the main.c, to move driver-related
functions under the drivers' source files.

Signed-off-by: George Pappas <pappasgeorge12@gmail.com>

xristosioan reviewed (approved):
gtg
